plugins/plugin-docker/che-plugin-docker-client/src/main/java/org/eclipse/che/plugin/docker/client/DockerConnector.java

```diff
@@ -711,32 +869,46 @@ public InputStream getResource(String container, String sourcePath) throws IOExc
      * @throws IOException
      *         when problems occurs with docker api calls, or during file system operations
      * @apiNote this method implements 1.20 docker API and requires docker not less than 1.8 version
+     * @deprecated use {@link #putResource(PutResourceParams)} instead
      */
+    @Deprecated
     public void putResource(String container,
                             String targetPath,
                             InputStream sourceStream,
                             boolean noOverwriteDirNonDir) throws IOException {
+       putResource(PutResourceParams.from(container, targetPath)
+                                    .withSourceStream(sourceStream)
+                                    .withNoOverwriteDirNonDir(noOverwriteDirNonDir));
+    }
+
+    /**
+     * Puts files into specified container.
+     *
+     * @throws IOException
+     *         when problems occurs with docker api calls, or during file system operations
+     * @apiNote this method implements 1.20 docker API and requires docker not less than 1.8 version
+     */
+    public void putResource(final PutResourceParams params) throws IOException {
         File tarFile;
         long length;
-        try (InputStream sourceData = sourceStream) {
+        try (InputStream sourceData = params.getSourceStream()) {
+            // we save stream to file, because we have to know its length
             Path tarFilePath = Files.createTempFile("compressed-resources", ".tar");
             tarFile = tarFilePath.toFile();
             length = Files.copy(sourceData, tarFilePath, StandardCopyOption.REPLACE_EXISTING);
         }
 
-        List<Pair<String, ?>> headers = Arrays.asList(Pair.of("Content-Type", ExtMediaType.APPLICATION_X_TAR),
-                                                      Pair.of("Content-Length", length));
         try (InputStream tarStream = new BufferedInputStream(new FileInputStream(tarFile));
              DockerConnection connection = connectionFactory.openConnection(dockerDaemonUri)
                                                             .method("PUT")
-                                                            .path("/containers/" + container + "/archive")
-                                                            .query("path", targetPath)
-                                                            .query("noOverwriteDirNonDir", noOverwriteDirNonDir ? 0 : 1)
-                                                            .headers(headers)
+                                                            .path("/containers/" + params.getContainer() + "/archive")
+                                                            .query("path", params.getTargetPath())
+                                                            .header("Content-Type", ExtMediaType.APPLICATION_X_TAR)
+                                                            .header("Content-Length", length)
```

Comment:
@mmorhun according to [http spec](http://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.4) it is possible to send request without specifying content length if chunked encoding header is set. So I assume that we don't have to save this tar file on disk! Please can you check if it is true with Docker API.